freqtrade/freqai/freqai_interface.py

```diff
@@ -0,0 +1,155 @@
+
+import os
+import numpy as np
+import pandas as pd
+from pandas import DataFrame
+import shutil
+import gc
+from typing import Any, Dict, Optional, Tuple
+from abc import ABC
+from freqtrade.freqai.data_handler import DataHandler
+
+pd.options.mode.chained_assignment = None
+
+class IFreqaiModel(ABC):
+    """
+    Class containing all tools for training and prediction in the strategy.
+    User models should inherit from this class as shown in 
+    templates/ExamplePredictionModel.py where the user overrides
+    train(), predict(), fit(), and make_labels().
+    Author: Robert Caulk, rob.caulk@gmail.com
+    """
+
+    def __init__(self, config: Dict[str, Any]) -> None:
+
+        self.config = config
+        self.freqai_info = config['freqai']
+        self.data_split_parameters = config['freqai']['data_split_parameters']
+        self.model_training_parameters = config['freqai']['model_training_parameters']
+        self.feature_parameters = config['freqai']['feature_parameters']
+        self.full_path = (str(config['user_data_dir'])+
+                            "/models/"+self.freqai_info['full_timerange']+
+                            '-'+self.freqai_info['identifier'])
+        self.metadata = {}
+        self.data = {}
+        self.time_last_trained = None
+        self.current_time = None
+        self.model = None
+        self.predictions = None
+
+        if not os.path.exists(self.full_path):
+            os.mkdir(self.full_path)
+            shutil.copy(self.config['config_files'][0],self.full_path+"/"+self.config['config_files'][0])
+
+    def start(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
+        """
+        Entry point to the FreqaiModel, it will train a new model if 
+        necesssary before making the prediction.
+        The backtesting and training paradigm is a sliding training window
+        with a following backtest window. Both windows slide according to the
+        length of the backtest window. This function is not intended to be 
+        overridden by children of IFreqaiModel, but technically, it can be 
+        if the user wishes to make deeper changes to the sliding window
+        logic.
+        :params:
+        :dataframe: Full dataframe coming from strategy - it contains entire
+        backtesting timerange + additional historical data necessary to train 
+        the model.
+        :metadata: pair metadataa coming from strategy. 
+        """
+        self.pair = metadata['pair']
+        self.dh = DataHandler(self.config, dataframe, self.data)
+
+        print('going to train',len(self.dh.training_timeranges),
+            'timeranges:',self.dh.training_timeranges)
```

Comment:
please avoid using print statements except for "user summary" output (like the backtesting tables).

If it's worthy output (which should be shown to the user) - please use `logger.info()` (or some other appropriate loglevel).
If it's just internal debug output while testing - i'd remove it completely (or comment it at least).